src/extension.rs

```diff
@@ -324,31 +324,39 @@ pub enum ExtensionBuildError {
 
 /// A set of extensions identified by their unique [`ExtensionId`].
 #[derive(Clone, Debug, Default, PartialEq, Eq, serde::Serialize, serde::Deserialize)]
-pub struct ExtensionSet(HashSet<ExtensionId>);
+pub struct ExtensionSet(HashSet<ExtensionId>, HashSet<usize>);
 
 impl ExtensionSet {
     /// Creates a new empty extension set.
     pub fn new() -> Self {
-        Self(HashSet::new())
+        Self(HashSet::new(), HashSet::new())
     }
 
     /// Creates a new extension set from some extensions.
     pub fn new_from_extensions(extensions: impl Into<HashSet<ExtensionId>>) -> Self {
-        Self(extensions.into())
+        Self(extensions.into(), HashSet::new())
     }
 
     /// Adds a extension to the set.
     pub fn insert(&mut self, extension: &ExtensionId) {
         self.0.insert(extension.clone());
     }
 
+    /// Adds a type var (which must have been declared as a [TypeParam::Extensions]) to this set
+    pub fn insert_type_var(&mut self, idx: usize) {
+        self.1.insert(idx);
+    }
+
     /// Returns `true` if the set contains the given extension.
     pub fn contains(&self, extension: &ExtensionId) -> bool {
         self.0.contains(extension)
     }
 
     /// Returns `true` if the set is a subset of `other`.
     pub fn is_subset(&self, other: &Self) -> bool {
+        if !other.1.is_empty() {
```

Comment:
the number of these panics suggests there are two different impls here squeezed in to one.

Reply:
Gone with malformed ExtensionIds instead...hackier but less intrusive